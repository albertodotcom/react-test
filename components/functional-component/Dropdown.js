import React from 'react';
import PropTypes from 'prop-types';

const Dropdown = props => (
  <select
    className="t-select"
    value={props.selected}
    onChange={e => props.onChange(e.target.value)}
  >
    { props.options.map(o =>
      <option className="t-option" key={o.id} value={o.id}>{o.text}</option>)}
  </select>
);

Dropdown.propTypes = {
  options: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.string.isRequired,
    text: PropTypes.string.isRequired,
  })).isRequired,
  selected: PropTypes.string,
  onChange: PropTypes.func.isRequired,
};

Dropdown.defaultProps = {
  selected: '',
};

export default Dropdown;
