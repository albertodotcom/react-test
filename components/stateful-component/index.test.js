/* eslint-env jest */
import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import Dropdown from './index';

describe('Dropdown', () => {
  const defaultProps = {
    options: [
      { id: '8', text: 'Conte' },
      { id: '10', text: 'Del Piero' },
      { id: '1', text: 'Buffon' },
    ],
    onSubmit: () => { },
  };

  describe('element testing', () => {
    it('renders the passed in options', () => {
      const options = [
        { id: '10', text: 'Del Piero' },
        { id: '21', text: 'Dybala' },
      ];

      const component = shallow(
        <Dropdown
          {...defaultProps}
          options={options}
        />);

      const optionTags = component.find('.t-option');
      expect(optionTags.length).toBe(2);
      expect(optionTags.at(0).text()).toEqual(options[0].text);
      expect(optionTags.at(1).text()).toEqual(options[1].text);
      expect(component.find('.t-selected-options').text()).toBe('');
    });

    it('renders the selected option', () => {
      const component = shallow(
        <Dropdown
          {...defaultProps}
          selected={['10']}
        />);

      expect(component.find('.t-selected-options').text()).toBe('Del Piero');

      component.find('select').simulate('change', { target: { value: '1' } });
      expect(component.find('.t-selected-options').text()).toBe('Del Piero, Buffon');
    });

    describe('DONTS', () => {
      it('displays the selected items by changing state', () => {
        const component = shallow(
          <Dropdown
            {...defaultProps}
            selected={['10']}
          />);

        expect(component.find('.t-selected-options').text()).toBe('Del Piero');

        component.setState({ selected: ['1', '10'] });
        expect(component.find('.t-selected-options').text()).toBe('Buffon, Del Piero');
      });
    });
  });

  describe('snapshot testing', () => {
    it('user can select options', () => {
      const onSubmit = jest.fn();
      const options = [
        { id: '8', text: 'Conte' },
        { id: '10', text: 'Del Piero' },
        { id: '1', text: 'Buffon' },
        { id: '21', text: 'Pirlo' },
      ];

      const wrapper = shallow(
        <Dropdown
          {...defaultProps}
          options={options}
          selected={['1', '21']}
          onSubmit={onSubmit}
        />);

      expect(toJson(wrapper)).toMatchSnapshot();

      wrapper.find('.t-select').simulate('change', { target: { value: '10' } });
      expect(toJson(wrapper)).toMatchSnapshot();

      wrapper.find('.t-submit').simulate('click');

      expect(onSubmit.mock.calls.length).toEqual(1);
      expect(onSubmit).toBeCalledWith(['1', '21', '10']);

      wrapper.find('.t-reset').simulate('click');
      expect(toJson(wrapper)).toMatchSnapshot();
    });
  });


  describe('iteractions', () => {
    it('clears the selection when clicking on reset', () => {
      const wrapper = shallow(
        <Dropdown
          {...defaultProps}
          selected={['10']}
        />);

      wrapper.find('.t-reset').simulate('click');

      expect(wrapper.find('.t-selected-options').text()).toEqual('');
    });

    it('returns the new selection on submit', () => {
      const onSubmit = jest.fn();
      const wrapper = shallow(
        <Dropdown
          {...defaultProps}
          selected={['10']}
          onSubmit={onSubmit}
        />);

      wrapper.find('.t-submit').simulate('click');

      expect(onSubmit.mock.calls.length).toEqual(1);
      expect(onSubmit).toBeCalledWith(['10']);
    });
  });
});
