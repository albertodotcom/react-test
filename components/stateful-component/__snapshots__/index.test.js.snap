// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown snapshot testing user can select options 1`] = `
<form>
  <div
    className="t-selected-options"
  >
    Buffon, Pirlo
  </div>
  <select
    className="t-select"
    onChange={[Function]}
  >
    <option
      className="t-option"
      value="8"
    >
      Conte
    </option>
    <option
      className="t-option"
      value="10"
    >
      Del Piero
    </option>
    <option
      className="t-option"
      value="1"
    >
      Buffon
    </option>
    <option
      className="t-option"
      value="21"
    >
      Pirlo
    </option>
  </select>
  <div>
    <button
      className="t-reset"
      onClick={[Function]}
      type="reset"
    >
      Clear
    </button>
    <button
      className="t-submit"
      onClick={[Function]}
      type="submit"
    >
      Submit
    </button>
  </div>
</form>
`;

exports[`Dropdown snapshot testing user can select options 2`] = `
<form>
  <div
    className="t-selected-options"
  >
    Buffon, Pirlo, Del Piero
  </div>
  <select
    className="t-select"
    onChange={[Function]}
  >
    <option
      className="t-option"
      value="8"
    >
      Conte
    </option>
    <option
      className="t-option"
      value="10"
    >
      Del Piero
    </option>
    <option
      className="t-option"
      value="1"
    >
      Buffon
    </option>
    <option
      className="t-option"
      value="21"
    >
      Pirlo
    </option>
  </select>
  <div>
    <button
      className="t-reset"
      onClick={[Function]}
      type="reset"
    >
      Clear
    </button>
    <button
      className="t-submit"
      onClick={[Function]}
      type="submit"
    >
      Submit
    </button>
  </div>
</form>
`;

exports[`Dropdown snapshot testing user can select options 3`] = `
<form>
  <div
    className="t-selected-options"
  />
  <select
    className="t-select"
    onChange={[Function]}
  >
    <option
      className="t-option"
      value="8"
    >
      Conte
    </option>
    <option
      className="t-option"
      value="10"
    >
      Del Piero
    </option>
    <option
      className="t-option"
      value="1"
    >
      Buffon
    </option>
    <option
      className="t-option"
      value="21"
    >
      Pirlo
    </option>
  </select>
  <div>
    <button
      className="t-reset"
      onClick={[Function]}
      type="reset"
    >
      Clear
    </button>
    <button
      className="t-submit"
      onClick={[Function]}
      type="submit"
    >
      Submit
    </button>
  </div>
</form>
`;
